### PR TITLE
Avoid duplicates ip address sent by the dhcp

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Adjust the code to avoid duplicate ip addresses given by the dhcp server

# Impacts
DHCP server

# Delete branch after merge
YES

# NEWS file entries

## Bug Fixes
* Issue with the dhcp server that give sometimes a duplicate ip addresse

### CODE REVIEW

File: go/dhcp/main.go [hunk 1]
```diff
@@ -16,14 +15,17 @@ import (
 	"time"
 
 	"github.com/coreos/go-systemd/daemon"
+	"github.com/davecgh/go-spew/spew"
```
Comment (satkunas) [inline go/dhcp/main.go:17]: Remove debugging

File: go/dhcp/main.go [hunk 2]
```diff
@@ -608,3 +673,11 @@ func (h *Interface) ServeDHCP(ctx context.Context, p dhcp.Packet, msgType dhcp.M
 	return answer
 
 }
+
+func recoverName(options dhcp.Options) {
+	if r := recover(); r != nil {
+		fmt.Println("recovered from ", r)
+		fmt.Println(errors.Wrap(r, 2).ErrorStack())
+		spew.Dump(options)
```
Comment (satkunas) [inline go/dhcp/main.go:690]: Is debugging needed here?

File: go/dhcp/pool/pool_test.go [hunk 3]
```diff
@@ -1,62 +1,61 @@
...
 
 	// Try to reserve an IP outside the capacity
-	err = dp.ReserveIPIndex(cap)
+	err, _ = dp.ReserveIPIndex(cap, mac)
 
 	if err == nil {
 		t.Error("Didn't get an error when trying to reserve an IP outside the capacity")
 	}
 }
-func TestIsFreeIPAtIndex(t *testing.T) {
```
Comment (julsemaan) [inline go/dhcp/pool/pool_test.go:37]: Why deleting the unit test ?

File: go/timedlock/rw_lock.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
+package timedlock
```
Comment (julsemaan) [inline go/timedlock/rw_lock.go]: Lets see if we can get #3255 merged with this change and have you rebase this on devel instead

File: go/dhcp/main.go (hunk 2 above)
Comment (fdurand) [reply go/dhcp/main.go:690]: It's a way to see which dhcp packet cause the crash.

File: go/dhcp/pool/pool_test.go (hunk 3 above)
Comment (fdurand) [reply go/dhcp/pool/pool_test.go:37]: I have removed the function IsFreeIPAtIndex

Comment (julsemaan) [reply go/dhcp/pool/pool_test.go:37]: It sounds like a pretty legit reason

**Comment by fdurand** [comment]
Rebased since  #3255 has been merged

File: go/dhcp/main.go [hunk 5]
```diff
@@ -154,9 +156,11 @@ func main() {
 	http.Handle("/", httpauth.SimpleBasicAuth(webservices.User, webservices.Pass)(router))
 
 	srv := &http.Server{
-		Addr:        "127.0.0.1:22222",
-		IdleTimeout: 5 * time.Second,
-		Handler:     router,
+		Addr:         "127.0.0.1:22222",
+		IdleTimeout:  5 * time.Second,
+		ReadTimeout:  5 * time.Second,
```
Comment (julsemaan) [inline go/dhcp/main.go]: Do the new read and write timeout really belong in this PR ?

File: go/dhcp/pool/pool.go [hunk 6]
```diff
@@ -6,6 +6,9 @@ import (
 	"sync"
 )
 
+const FreeMac = "00:00:00:00:00:00"
```
Comment (julsemaan) [inline go/dhcp/pool/pool.go:9]: Create a package for constants and centralize them there so they aren't duplicated in each package

**Pull request merged**